flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/table/KuduTableFactory.java

```diff
@@ -0,0 +1,169 @@
+import static org.apache.flink.table.descriptors.Rowtime.ROWTIME_WATERMARKS_TYPE;
+import static org.apache.flink.table.descriptors.Schema.SCHEMA;
+import static org.apache.flink.table.descriptors.Schema.SCHEMA_DATA_TYPE;
+import static org.apache.flink.table.descriptors.Schema.SCHEMA_FROM;
+import static org.apache.flink.table.descriptors.Schema.SCHEMA_NAME;
+import static org.apache.flink.table.descriptors.Schema.SCHEMA_PROCTIME;
+import static org.apache.flink.table.descriptors.Schema.SCHEMA_TYPE;
+
+public class KuduTableFactory implements TableSourceFactory<Row>, TableSinkFactory<Tuple2<Boolean, Row>> {
+
+    public static final String KUDU_TABLE = "kudu.table";
+    public static final String KUDU_MASTERS = "kudu.masters";
+    public static final String KUDU_HASH_COLS = "kudu.hash-columns";
+    public static final String KUDU_PRIMARY_KEY_COLS = "kudu.primary-key-columns";
+    public static final String KUDU_REPLICAS = "kudu.replicas";
+    public static final String KUDU = "kudu";
+
+    @Override
+    public Map<String, String> requiredContext() {
+        Map<String, String> context = new HashMap<>();
+        context.put(CONNECTOR_TYPE, KUDU);
+        return context;
+    }
+
+    @Override
+    public List<String> supportedProperties() {
+        List<String> properties = new ArrayList<>();
+        properties.add(KUDU_TABLE);
+        properties.add(KUDU_MASTERS);
+        properties.add(KUDU_HASH_COLS);
+        properties.add(KUDU_PRIMARY_KEY_COLS);
+        // schema
+        properties.add(SCHEMA + ".#." + SCHEMA_DATA_TYPE);
+        properties.add(SCHEMA + ".#." + SCHEMA_TYPE);
+        properties.add(SCHEMA + ".#." + SCHEMA_NAME);
+        properties.add(SCHEMA + ".#." + SCHEMA_FROM);
+        // computed column
+        properties.add(SCHEMA + ".#." + TABLE_SCHEMA_EXPR);
+
+        // time attributes
+        properties.add(SCHEMA + ".#." + SCHEMA_PROCTIME);
+        properties.add(SCHEMA + ".#." + ROWTIME_TIMESTAMPS_TYPE);
+        properties.add(SCHEMA + ".#." + ROWTIME_TIMESTAMPS_FROM);
+        properties.add(SCHEMA + ".#." + ROWTIME_TIMESTAMPS_CLASS);
+        properties.add(SCHEMA + ".#." + ROWTIME_TIMESTAMPS_SERIALIZED);
+        properties.add(SCHEMA + ".#." + ROWTIME_WATERMARKS_TYPE);
+        properties.add(SCHEMA + ".#." + ROWTIME_WATERMARKS_CLASS);
+        properties.add(SCHEMA + ".#." + ROWTIME_WATERMARKS_SERIALIZED);
+        properties.add(SCHEMA + ".#." + ROWTIME_WATERMARKS_DELAY);
+
+        // watermark
+        properties.add(SCHEMA + "." + WATERMARK + ".#." + WATERMARK_ROWTIME);
+        properties.add(SCHEMA + "." + WATERMARK + ".#." + WATERMARK_STRATEGY_EXPR);
+        properties.add(SCHEMA + "." + WATERMARK + ".#." + WATERMARK_STRATEGY_DATA_TYPE);
+        return properties;
+    }
+
+    private DescriptorProperties getValidatedProps(Map<String, String> properties) {
+        checkNotNull(properties.get(KUDU_MASTERS), "Missing required property " + KUDU_MASTERS);
```

Comment:
Other integrations have provided an application level configuration to use a default if a per-table override is not provided. Does that make sense here too?

Reply:
The KuduCatalog basically provides the default for this. So as long as you use the KuduCatalog (and specified the masters there) you wouldnt have to specify it for individual tables. 

On the other hand when you are creating Kudu tables in other catalogs you will end up in this codepath and you have to specify the kudu masters